Add Help and Resources

ax2mx opened the pull request:
- Add Help Menu 
- Add Shortcuts Panel
- Add optional launching developer server link in browser
- Fix double click on Editor in edit mode
- Add basic node insertion on click
- Prevent default Select All behavior
- Prevent caret jumping on value change
- Add node insertion via arrow keys
- Add undoable range comment
- Prevent insertion of initial circular dependency node
- Add cell content preview on hover

ax2mx committed:
Remove showing NodeNavigator on:mouseenter

ax2mx committed:
Fix NodeNavigator opacity preventing hide

ax2mx committed:
Add help menu

ax2mx committed:
Merge branch 'dmaevsky:master' into master

ax2mx committed:
Fix Navigator and Shortcuts menu item

Now these menu items open Node Navigator and Keyboard shortcuts
helper panels respectively

ax2mx committed:
Fix closing button bug

ax2mx committed:
Add Keyboard shortcut panel

- Set basic layout and styling
- Add Shortcut component to simplify layout

ax2mx committed:
Add Shortcuts panel

- Add SortcutsHelper component
- Add keyboard combinations to open Shortcuts and Node Navigator

ax2mx committed:
Add dark mode support for shortcuts

ax2mx committed:
Fix shortcuts description

ax2mx committed:
Make shortcuts panel responsive

ax2mx committed:
Fix Help menu and shotrcut styles

ax2mx committed:
Add top border

ax2mx committed:
Update UI

ax2mx committed:
Fix z-index

ax2mx committed:
Fix help layout and styles

ax2mx committed:
Detect slash key

ax2mx committed:
Fix unclickable sheet, add comment shortcut

ax2mx committed:
Auto open localhost in browser

ax2mx committed:
Fix double click in edit mode

ax2mx committed:
Add option to open devserver link in browser

ax2mx committed:
Get rid of clickOutside()

ax2mx committed:
Add Edit mode detection

- Detect Edit mode via binding isEditMode prop in Grid.svelte
- Prevent default Select All behavior
- Simplify comma operator expressions for better readability
- Slight code restyling

ax2mx committed:
Handle Editor mode

- Detect formula in Editor
- Prevent default Select All behavior in Edit mode
- Handle Editor clicks and keystrokes
- Exit formula input on double click
- Slight refactoring

ax2mx committed:
Switch to Edit mode on Enter pressed

ax2mx committed:
Set default cursor on node insert, small restyling

ax2mx committed:
Change active border color, small restyling

ax2mx committed:
Add basic node insertion in formula

- Add onclick node insertion in current formula
- Add getCaretPosition()
- Some code restyling to enhance readability

ax2mx committed:
Prevent caret jumping on value change

ax2mx committed:
Use tick() instead of requestAnimationFrame()

ax2mx committed:
Reset caret position on closing editor

ax2mx committed:
Get caret position on click

ax2mx committed:
Fix simple formula detection

ax2mx committed:
Add source node highlighting during insertion

ax2mx committed:
Simplify node insert handling on click

ax2mx committed:
Add basic node insertion via keyboard

ax2mx committed:
Add extended node insertion support

ax2mx committed:
Regroup code

ax2mx committed:
Add scroll into view for highlight

ax2mx committed:
Fix highlight/selection style overlap

ax2mx committed:
Prevent select on jump away

ax2mx committed:
Simplify highlight variable

ax2mx committed:
Add reusable getCoords function

ax2mx committed:
Add undoable range comments

ax2mx committed:
Get rid of unused props in CellEditor

dmaevsky commented on src/index.js:
```diff
@@ -39,31 +41,39 @@ if (mainOptions.command === 'start') {
+      const url = `http://localhost:${config.port}`;
+
+      console.log(`> Running on ${url}`);
+      if (config.open) await open(url);
```
why do you need await?

ax2mx replied:
Actually, I don't in this very situation. It can be safely removed as we don't use a promise.

This is what [API Docs](https://www.npmjs.com/package/open) says:

> `open(url)` returns a promise for the spawned child process. You would normally not need to use this for anything, but it can be useful if you'd like to attach custom event listeners or perform other operations directly on the spawned process.

ax2mx replied:
Done.

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -13,33 +15,63 @@
+    insertRange = null;
+    isArrowMode = false;
+  }
+  $: if (isEditMode) highlight = selection;
```
"else" would be better

ax2mx replied:
Sorry, but I'm not quite sure I understand well the meaning of this comment. Could you please explain it?

dmaevsky replied:
You have 2 lines:
`$: if (!isEditMode)`and
`$: if (isEditMode) `

and then
`$: if (isArrowMode)`and
`$: if (!isArrowMode)`
`if () else` would do, no?

ax2mx replied:
Yes, that would be the way better. Done in upcoming commit!

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -13,33 +15,63 @@
+  $: if (isEditMode) highlight = selection;
+
+  $: if (isArrowMode) [arrowRow, arrowCol] = highlight ? highlight : selection;
+  $: if (!isArrowMode) highlight = selection;
```
isn't this redundant?

ax2mx replied:
Yes. This logic is simplified in upcoming commit.

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -61,179 +92,138 @@
       width: ${(colEnd - colStart + 1) * columnWidth + 2}px;
     `
   })(copySelection);
+  $: if (selection) requestAnimationFrame(() => scrollIntoView(selection));
+  $: if (highlight) requestAnimationFrame(() => scrollIntoView(highlight));
```
I can imagine the first statement should not execute if highlight is present. So might be if (highlight) {} else {} is better?

ax2mx replied:
Perhaps this would be even better:

```suggestion
  $: requestAnimationFrame(() => scrollIntoView(isEditMode? highlight : selection));
```

Implemented in upcoming commits.

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -61,179 +92,138 @@
+  function getValue(row, col) {
+    let blockId = query(blocks).getAt(row, col);
+    let block = blocks.get(blockId);
+    if (blockId !== undefined && block.formula !== undefined) {
+      return block.node;
+    }
```
I believe formula can be there while node is not: in the situation where there's a parsing error

ax2mx replied:
Fixed!

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -61,179 +92,138 @@
-    if (left < scrollLeft) container.scrollLeft = left - 0.5 * clientWidth;
-    else if (right > scrollLeft + clientWidth) container.scrollLeft = right - clientWidth;
-  }
+          [start, end] = isNodeInserted
```
let [start, end] = ?

ax2mx replied:
Fixed.

dmaevsky commented on src/sandbox/runtime/actions/edit.js:
```diff
@@ -55,3 +70,20 @@ export const clearRange = undoable((thisSheet, range) => {
   for (let blockId of found) clearBlock(thisSheet, blockId);
 });
+
+export const commentRange = undoable((thisSheet, range, cb) => {
```
It's very unclear what cb is for. Can it be anything but just toggling "//" ?

ax2mx replied:
I agree. There's actually no other possible functionality here except toggling comments.
Fixed.

dmaevsky commented on src/sandbox/runtime/actions/edit.js:
```diff
@@ -55,3 +70,20 @@ export const clearRange = undoable((thisSheet, range) => {
+  for (let blockId of found) {
+    const block = blocks.get(blockId);
+    const [row, col] = block.position;
+    const value = block
```
If block is undefined block.position would have thrown the line before

ax2mx replied:
Could you please provide an example?

ax2mx replied:
OK. Fixed!

dmaevsky replied:
How is this fixed??? Renamed the callback??? DRY should not mean spaghetti!
toggleComment is just a utility function that you need both in edit.js and in the component. Then wtf is the reason for it to remain in the component???

ax2mx committed:
Fix scrollIntoView for highlight

ax2mx committed:
Simplify arrow mode highlight logic

ax2mx committed:
Simplify scrollIntoView logic

ax2mx committed:
Get rid of await

ax2mx committed:
Add more node checks before insertion

- Prevent insertion of initial circular dependency node
- Check block.node existence

ax2mx committed:
Small refactoring

ax2mx committed:
Rename action cb to toggleComment

ax2mx committed:
Add cell content preview on hover

ax2mx committed:
Remove useless check

dmaevsky commented on src/sandbox/components/GridLayout.svelte:
```diff
@@ -119,19 +127,70 @@
+          class:bg-transparent={transparent}
+          class:header={header}
+          class="gridlayout__cell gridlines"
+          title={showContent(row, col)}
```
This is ugly as hell. Can you even think of the execution complexity here ??? It's at least smth like n^4 and probably worse.
I believe on a sheet with a million cells this is gonna die a hard death.

ax2mx committed:
Get rid of computational hell

ax2mx committed:
Get rid of redundant isNodeInserted variable

ax2mx committed:
Move toggleComment to edit.js

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -61,179 +93,125 @@
       width: ${(colEnd - colStart + 1) * columnWidth + 2}px;
     `
   })(copySelection);
+  $: requestAnimationFrame(() => scrollIntoView(isEditMode? highlight : selection));
```
This will not be reactive. You removed `if (selection)` that was there before for this purpose

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -13,33 +14,65 @@
+
+  $: if (isArrowMode) [arrowRow, arrowCol] = highlight;
+
+  $: console.log("~ insertRange: ", insertRange )
```
remove `console.log`

ax2mx replied:
Done.

ax2mx committed:
DRY setInsertRange()

ax2mx committed:
DRY handleArrows
